Support opening/closing luks encrypted ceph volumes

axqx opened the pull request:
Solves https://jira.medallia.com/browse/CIF-30

Tests:

1. Test mount ceph encrypted volume using docker

1.1. Manually create encrypted volume
```
rbd create lala-1 --size 16G
 sudo rbd map lala-1 
sudo cryptsetup luksFormat /dev/rbd0 # lala-1
cryptsetup luksOpen /dev/rbd0 lala-1 
mkfs.ext4 /dev/mapper/lala-1
mkdir /vagrant/data/lala-1 
mount /dev/mapper/lala-1 /vagrant/data/lala-1
echo "dog" > /vagrant/data/lala-1/cat
```

1.2. Mount volume with docker container
```
medallia@fib-r10-u08:~$ docker run -ti -v lala-1:/foo:ceph,rw ubuntu /bin/bash
root@90f91393252d:/# cat /foo/cat 
dog
```

2. Test can't mount same ceph encrypted volume with docker twice

2.1 Try to run second container while one is still running
```
medallia@fib-r10-u01:~$ docker run -ti -v lala-1:/foo:ceph,rw ubuntu /bin/bash
docker: Error response from daemon: Failed to map Ceph volume 'lala-1': exit status 1 - (2460) Map lala-1
(2460) Map lala-1 failed: image is being watched: watcher=10.112.10.34:0/1093050966 client.114105 cookie=8.
```

2.2 kill first container and then re-run

```
medallia@fib-r10-u01:~$ docker run -ti -v lala-1:/foo:ceph,rw ubuntu /bin/bash
root@ef23c13041e8:/# cat /foo/cat 
dog
```

3. hard power-cycle machine while docker container with encrypted volume is running and try to re-mount volume on docker container in another machine 

This works in FIB but I had to manually remove locks and watchers for the volume since FIB is not doing this automatically. Will test in Germany.

Commit:
Support opening/closing luks encrypted ceph volumes

thorvald commented on volume/ceph/ceph.go:
```diff
@@ -15,7 +17,11 @@ import (
+const (
+	CephImageSizeMB   = 1024 * 1024 // 1TB
+	LuksDevMapperPath = "/dev/mapper/"
+	criptoLUKS        = "crypto_LUKS"
```
Typo: Crypto

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/a10170555e5c350567eb59e9539beadb59721237.

thorvald commented on volume/ceph/ceph.go:
```diff
@@ -177,22 +186,56 @@ func (v *Volume) Mount(id string) (mappedDevicePath string, returnedError error)
 	}
+
+	if fsType == criptoLUKS {
+		cmd = exec.Command("cryptsetup", "luksOpen", v.mappedDevicePath, v.Name())
```
Let's be nice to cryptsetup, and pass "-q" so it knows we're in batch mode. You might also want to pass "--key-file=-" to explicitly get the key from stdin in a programmatic way.

axqx replied:
Added "--key-file=-" in https://github.com/medallia/docker/pull/15/commits/a10170555e5c350567eb59e9539beadb59721237

 --batch-mode is not relevant for luksOpen:

```
--batch-mode, -q
Do not ask for confirmation. Use with care! This option is only relevant for luksFormat, luksAddKey, luksRemoveKey or luksKillSlot.
```

thorvald commented on volume/ceph/ceph.go:
```diff
@@ -177,22 +186,56 @@ func (v *Volume) Mount(id string) (mappedDevicePath string, returnedError error)
+		v.mappedLuksDevicePath = path.Join(LuksDevMapperPath, v.Name())
+	}
+
 	if fsType == "" {
```
Before we get here, we need to recheck the fsType, but now checking it from the opened volume. We do want to run fsck etc on the opened cryptovolume.

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/a10170555e5c350567eb59e9539beadb59721237

Commit:
Address comments from Thorvald

axqx commented:
@thorvald, thank you for the comments. Updates added in https://github.com/medallia/docker/pull/15/commits/a10170555e5c350567eb59e9539beadb59721237

mauri commented on volume/ceph/ceph.go:
```diff
@@ -177,24 +186,64 @@ func (v *Volume) Mount(id string) (mappedDevicePath string, returnedError error)
 	if fsType == "" {
-		cmd = exec.Command("mkfs.ext4", "-m0", "-E", "nodiscard,lazy_itable_init=0,lazy_journal_init=0,packed_meta_blocks=1", v.mappedDevicePath)
-		logrus.Infof("Creating ext4 filesystem in newly created Ceph volume '%s' (device %s)", v.Name(), v.mappedDevicePath)
+		//if fsType == "" {
```
remove?

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/975c9c2ab2e543bd69892b3254de4e3e90c1156e

mauri commented on volume/ceph/ceph.go:
```diff
@@ -204,6 +253,19 @@ func (v *Volume) Unmount(id string) error {
+		return err
+	}
+
+	if fsType == "crypto_LUKS" {
```
use the constant you extracted

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/975c9c2ab2e543bd69892b3254de4e3e90c1156e

mauri commented on volume/ceph/ceph.go:
```diff
@@ -177,24 +186,64 @@ func (v *Volume) Mount(id string) (mappedDevicePath string, returnedError error)
 		return "", err
 	}
+
+	mappedDevicePath = v.mappedDevicePath
```
rename to deviceToMount to avoid confusions with v.mappedDevicePath?

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/975c9c2ab2e543bd69892b3254de4e3e90c1156e.

mauri commented on volume/ceph/ceph.go:
```diff
@@ -15,7 +17,11 @@ import (
+const (
+	CephImageSizeMB   = 1024 * 1024 // 1TB
+	LuksDevMapperPath = "/dev/mapper/"
+	cryptoLUKS        = "crypto_LUKS"
```
cryptoLuksFsType? I think consts uses Uppercase?

axqx replied:
Fixed in https://github.com/medallia/docker/pull/15/commits/975c9c2ab2e543bd69892b3254de4e3e90c1156e

mauri commented on volume/ceph/ceph.go:
```diff
@@ -177,24 +186,64 @@ func (v *Volume) Mount(id string) (mappedDevicePath string, returnedError error)
+		}
+	}
+
 	if fsType == "" {
```
else if?

thorvald replied:
We want a straight if; If the image is luks encrypted, but does NOT have a filesystem, we want to create a filesystem after we lukeOpen the image.

mauri replied:
didn't see the overriding fsType inside the if block

thorvald commented:
My requested changes look good; address Mauri's too and we should be GTG.

Commit:
Address Mauricio comments

axqx commented:
@mauri commit https://github.com/medallia/docker/pull/15/commits/975c9c2ab2e543bd69892b3254de4e3e90c1156e addressed your comments. If you don't have additional comments and this looks good to you could you merge the PR please?
